### PR TITLE
[DO NOT MERGE] Pagination Tutorial

### PR DESCRIPTION
**Pull request by ramyanaga**

### CODE REVIEW

File: client/src/App.css [hunk 1]
```diff
@@ -59,7 +59,7 @@ div.NotFound {
 
 div.channelsList {
   padding: 0 30px;
-  position: fixed;
+   position: fixed;
```
Comment (stubailo) [inline client/src/App.css]: wat

File: client/src/components/AddMessage.js [hunk 2]
```diff
@@ -30,15 +30,16 @@ const AddMessage = ({ mutate, match }) => {
...
             // Add our Message from the mutation to the end.
-            data.channel.messages.push(addMessage);
+            data.channel.messageFeed.messages.push(addMessage)
           }
           // Write the data back to the cache.
           store.writeQuery({
             query: channelDetailsQuery,
             variables: {
               channelId: match.params.channelId,
+
```
Comment (stubailo) [inline client/src/components/AddMessage.js]: rip

File: client/src/components/ChannelDetails.js [hunk 3]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+    variables: { 
+      channelId: props.match.params.channelId,
+
+    },
   }),
+
+  props: (props) => {
+
+    return Object.assign(props, {
+      loadMore: () => {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Can we name this something more descriptive, like `loadOlderMessages`

File: client/src/components/ChannelDetails.js [hunk 4]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
 
 export default (graphql(channelDetailsQuery, {
   options: (props) => ({
-    variables: { channelId: props.match.params.channelId },
+    variables: { 
+      channelId: props.match.params.channelId,
+
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: empty line

File: client/src/components/ChannelDetails.js [hunk 5]
```diff
@@ -48,32 +46,40 @@ class ChannelDetails extends Component {
...
+          text
+        }
       }
     }
   }
 `;
 
 const messagesSubscription = gql`
-  subscription messageAdded($channelId: ID!) {
+  subscription messageAdded($channelId: Int) {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Why this change?

File: client/src/components/ChannelDetails.js [hunk 6]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
-    variables: { channelId: props.match.params.channelId },
+    variables: { 
+      channelId: props.match.params.channelId,
+
+    },
   }),
+
+  props: (props) => {
+
+    return Object.assign(props, {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: No need to `assign` here - you can just return the object you were assigning

File: client/src/components/ChannelDetails.js [hunk 7]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+      loadMore: () => {
+        return props.data.fetchMore({
+          variables: {
+            //channelId: props.match.params.channelId,
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: comment

File: client/src/components/ChannelDetails.js [hunk 8]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+            channelId: props.data.channel.id ? props.data.channel.id : props.match.params.channelId,
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: any reason not to just always use `props.match.params.channelId`?

File: client/src/components/ChannelDetails.js [hunk 9]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+            //channelId: props.match.params.channelId,
+            channelId: props.data.channel.id ? props.data.channel.id : props.match.params.channelId, 
+            cursor: props.data.channel.messageFeed.cursor
+          },
+          updateQuery(previousResult, { fetchMoreResult }) {
+            let prevMessageFeed = previousResult.channel.messageFeed
+            let newMessageFeed = fetchMoreResult.channel.messageFeed
+            const newChannelData = Object.assign({}, previousResult.channel, {
+                messageFeed: {
+                  messages: [...fetchMoreResult.channel.messageFeed.messages, ...previousResult.channel.messageFeed.messages],
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Use the vars you declared above

File: client/src/components/ChannelDetails.js [hunk 10]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+    return Object.assign(props, {
+      loadMore: () => {
+        return props.data.fetchMore({
+          variables: {
+            //channelId: props.match.params.channelId,
+            channelId: props.data.channel.id ? props.data.channel.id : props.match.params.channelId, 
+            cursor: props.data.channel.messageFeed.cursor
+          },
+          updateQuery(previousResult, { fetchMoreResult }) {
+            let prevMessageFeed = previousResult.channel.messageFeed
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: const

File: client/src/components/ChannelDetails.js [hunk 11]
```diff
@@ -83,6 +89,37 @@ const messagesSubscription = gql`
...
+            let newMessageFeed = fetchMoreResult.channel.messageFeed
+            const newChannelData = Object.assign({}, previousResult.channel, {
+                messageFeed: {
+                  messages: [...fetchMoreResult.channel.messageFeed.messages, ...previousResult.channel.messageFeed.messages],
+                  cursor: newMessageFeed.cursor
+                }
+            });
+
+            const newData = Object.assign({}, previousResult, { channel: newChannelData });
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Use `{ ...previousResult, { channel: newChannelData }}`

Basically let's always use object spread instead of assign

File: server/src/resolvers.js [hunk 12]
```diff
@@ -10,20 +10,38 @@ const channels = [{
...
+    text: 'message5'
+  }, {
+    id: '6',
+    text: 'message6'
+  }, {
+    id: '7',
+    text: 'message7'
+  }, {
+    id: '8',
+    text: 'message8'
```
Comment (stubailo) [inline server/src/resolvers.js]: Let's replace this with faker or something? And generate like 50 messages

File: server/src/resolvers.js [hunk 13]
```diff
@@ -32,8 +50,28 @@ export const resolvers = {
     channels: () => {
       return channels;
     },
-    channel: (root, { id }) => {
-      return channels.find(channel => channel.id === id);
+
+    channel: (root, args) => {
+      let id = args['id'];
```
Comment (stubailo) [inline server/src/resolvers.js]: `args.id` - no reason to use the bracket syntax here

File: server/src/resolvers.js [hunk 14]
```diff
@@ -32,8 +50,28 @@ export const resolvers = {
...
+      let cursor = args['cursor'];
+      let channel = channels.find(channel => channel.id === id);
+      if (cursor == undefined && messageFeed == undefined) {
```
Comment (stubailo) [inline server/src/resolvers.js]: what's `messageFeed` here?

Comment (stubailo) [inline server/src/resolvers.js]: I think it's always undefined.

This could be replaced with `cursor = args.cursor || channel.messages.length`

File: server/src/resolvers.js [hunk 15]
```diff
@@ -32,8 +50,28 @@ export const resolvers = {
...
+        cursor = channel.messages.length;
+      }
+
+      let messageFeed = {
+        messages: channel.messages.slice(cursor-2, cursor),
+        cursor: cursor - 2
```
Comment (stubailo) [inline server/src/resolvers.js]: Can we factor out the page size here? `2` is a magic number. I think we also want to make that an argument from the client, like `limit: 2` or whatevs. Start by factoring it into a var though

File: server/src/resolvers.js [hunk 16]
```diff
@@ -32,8 +50,28 @@ export const resolvers = {
...
+      }
+
+      channel.messageFeed = messageFeed;
```
Comment (stubailo) [inline server/src/resolvers.js]: looks like this is overwriting our "database" object - let's not do that

File: server/src/resolvers.js [hunk 17]
```diff
@@ -32,8 +50,28 @@ export const resolvers = {
...
+      let lessChannel = {
```
Comment (stubailo) [inline server/src/resolvers.js]: what's `lessChannel` mean?

File: client/src/components/ChannelDetails.js [hunk 18]
```diff
@@ -48,25 +46,33 @@ class ChannelDetails extends Component {
...
       <div>
+      <div>
         <div className="channelName">
           {channel.name}
         </div>
-        <MessageList messages={channel.messages}/>
-      </div>);
+        <MessageList messages={channel.messageFeed.messages}/>
+      </div>
+      <button onClick={loadOlderMessages}>
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Can we move this to the top and make it say "load older messages"?

File: client/src/components/ChannelDetails.js [hunk 19]
```diff
@@ -48,25 +46,33 @@ class ChannelDetails extends Component {
...
-  query ChannelDetailsQuery($channelId : ID!) {
-    channel(id: $channelId) {
+  query ChannelDetailsQuery($channelId : ID!, $cursor: Int, $limit: Int) {
+    channel(id: $channelId, cursor: $cursor, limit: $limit) {
       id
       name
-      messages {
-        id
-        text
+      messageFeed(cursor: $cursor, limit: $limit) @connection(key: "messageFeed") {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: I think we can put `limit` directly in here, doesn't need to be a variable

File: client/src/components/ChannelDetails.js [hunk 20]
```diff
@@ -83,6 +89,35 @@ const messagesSubscription = gql`
...
   options: (props) => ({
-    variables: { channelId: props.match.params.channelId },
+    variables: { 
+      channelId: props.match.params.channelId,
+      limit: 10
+    },
   }),
+
+  props: (props) => {
+    return Object.assign(props, {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Reminder - we don't need `Object.assign` here

File: client/src/components/ChannelDetails.js [hunk 21]
```diff
@@ -83,6 +89,35 @@ const messagesSubscription = gql`
...
+        return props.data.fetchMore({
+          variables: {
+            channelId: props.data.channel.id,
+            cursor: props.data.channel.messageFeed.cursor,
+            limit: 10
+          },
+          updateQuery(previousResult, { fetchMoreResult }) {
+            const prevMessageFeed = previousResult.channel.messageFeed
+            const newMessageFeed = fetchMoreResult.channel.messageFeed
+            const newChannelData = Object.assign({}, previousResult.channel, {
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: Same here - use `{...}`

File: server/src/resolvers.js [hunk 22]
```diff
@@ -10,20 +12,49 @@ const channels = [{
   }, {
     id: '2',
     text: 'hello soccer world cup',
+  }, {
+    id: '3',
```
Comment (stubailo) [inline server/src/resolvers.js]: Do we need these or can we just use faker for everything?

File: server/src/resolvers.js [hunk 23]
```diff
@@ -10,20 +12,49 @@ const channels = [{
...
   }]
 }];
+
+// use faker to generate random messages in soccer channel
+const channel = channels.find(channel => channel.name === 'soccer');
+let id;
+for (id = 9; id < 50; id++) {
+  channel.messages.push({
+    id: id,
+    text: faker.random.words()
```
Comment (stubailo) [inline server/src/resolvers.js]: Let's set a seed so we always get the same results: https://github.com/marak/Faker.js/#setting-a-randomness-seed

File: server/src/resolvers.js [hunk 24]
```diff
@@ -10,20 +12,49 @@ const channels = [{
...
+  });
+}
+
 let nextId = 3;
-let nextMessageId = 5;
+let nextMessageId = 52;
```
Comment (stubailo) [inline server/src/resolvers.js]: can this just be like `messages.length` or `messages[messages.length - 1].id`

File: server/src/resolvers.js [hunk 25]
```diff
@@ -32,8 +63,28 @@ export const resolvers = {
...
+      if (cursor == undefined && messageFeed == undefined) {
+        cursor = channel.messages.length;
+      }
+
+      let messageFeed = {
+        messages: channel.messages.slice(cursor-limit, cursor),
+        cursor: cursor - limit
+      }
+
+      let lessChannel = {
```
Comment (stubailo) [inline server/src/resolvers.js]: What does `lessChannel` mean?

File: client/src/components/ChannelDetails.js (hunk 19 above)
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: The API should have that option, but on the client we can just hardcode a consistent page size

File: client/src/components/ChannelDetails.js (hunk 8 above)
Comment (ramyanaga) [inline client/src/components/ChannelDetails.js]: it wasn't defined when I console.logged it, so I just changed it to always use props.data.channel.id

File: server/src/resolvers.js (hunk 24 above)
Comment (ramyanaga) [inline server/src/resolvers.js]: Can I get rid of the basketball channel for this tutorial and then nextMessageId can be channel.messages.length ?

File: client/src/components/ChannelDetails.js [hunk 26]
```diff
@@ -83,6 +89,33 @@ const messagesSubscription = gql`
...
 export default (graphql(channelDetailsQuery, {
   options: (props) => ({
-    variables: { channelId: props.match.params.channelId },
+    variables: { 
+      channelId: props.match.params.channelId,
+    },
   }),
+
+  props: (props) => {
+    return {...props,
```
Comment (stubailo) [inline client/src/components/ChannelDetails.js]: react apollo already passes through the parent props.

File: server/src/resolvers.js [hunk 27]
```diff
@@ -32,8 +42,27 @@ export const resolvers = {
     channels: () => {
       return channels;
     },
-    channel: (root, { id }) => {
-      return channels.find(channel => channel.id === id);
+
+    channel: (root, args) => {
```
Comment (stubailo) [inline server/src/resolvers.js]: Can we make this `(root, { id, cursor })` - instead of using `args.`

File: server/src/resolvers.js [hunk 28]
```diff
@@ -32,8 +42,27 @@ export const resolvers = {
...
+      let id = args.id
+      let cursor = args.cursor
+      let channel = channels.find(channel => channel.id === id);
+      if (cursor == undefined && messageFeed == undefined) {
```
Comment (stubailo) [inline server/src/resolvers.js]: Usually people do `if(!cursor)`. Also, `messageFeed` is _always_ going to be undefined here.

File: server/src/resolvers.js [hunk 29]
```diff
@@ -1,29 +1,39 @@
...
   }, {
-    id: '4',
+    id: '51',
     text: 'hello baseball world series',
   }]
 }];
-let nextId = 3;
-let nextMessageId = 5;
+
+// use faker to generate random messages in soccer channel
```
Comment (stubailo) [inline server/src/resolvers.js]: This comment is not correct

File: server/src/resolvers.js [hunk 30]
```diff
@@ -1,29 +1,39 @@
...
+const channel = channels.find(channel => channel.name === 'faker');
+let id;
+
+// Add seed for consistent random data
+faker.seed(9);
+for (id = 0; id < 50; id++) {
```
Comment (stubailo) [inline server/src/resolvers.js]: Put the `let` inside the `for( ... )`

File: server/src/resolvers.js [hunk 31]
```diff
@@ -1,29 +1,39 @@
...
+  channel.messages.push({
+    id: id,
+    text: faker.random.words()
+  });
+}
+
+let nextId = channels.length;
+let nextMessageId = 52;
```
Comment (stubailo) [inline server/src/resolvers.js]: Do we need to hardcode this here?

**Comment by stubailo** [comment]
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134356688
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134357242
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134357287
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134357342
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134356929
- [ ] https://github.com/apollographql/graphql-tutorial/pull/22#discussion_r134357095

**Comment by stubailo** [comment]
(This PR was for reviewing the code changes)
